packages/@react-aria/datepicker/src/useDateField.ts

```diff
@@ -126,7 +126,48 @@ export function useDateField<T extends DateValue>(props: AriaDateFieldProps<T>,
         focusManager.focusFirst();
       }
     },
-    fieldProps: mergeProps(domProps, fieldDOMProps, groupProps, focusWithinProps),
+    fieldProps: mergeProps(domProps, fieldDOMProps, groupProps, focusWithinProps, {
+      onKeyDown(e: KeyboardEvent) {
+        if (props.onKeyDown) {
+          props.onKeyDown(e);
+        }
+      },
+      onKeyUp(e: KeyboardEvent) {
+        if (props.onKeyUp) {
+          props.onKeyUp(e);
+        }
+      },
+      onFocus(e: ReactFocusEvent) {
+        if (state.isFocused) {
+          return;
+        }
+
+        if (props.onFocus) {
+          props.onFocus(e);
+        }
+
+        if (props.onFocusChange) {
+          props.onFocusChange(true);
+        }
+
+        state.setFocused(true);
+      },
+      onBlur(e: ReactFocusEvent) {
+        if (e.currentTarget.contains(e.relatedTarget as Node)) {
+          return;
+        }
+
+        if (props.onBlur) {
+          props.onBlur(e);
+        }
+
+        if (props.onFocusChange) {
+          props.onFocusChange(false);
+        }
+
+        state.setFocused(false);
+      }
```

Comment:
FYI, this is basically recreating the logic that useFocusWithin already implements. We could probably combine.